network/crl.go

```diff
@@ -0,0 +1,370 @@
+func NewLRUCRLCache(maxEntries int) *LRUCRLCache {
+	return &LRUCRLCache{cache: lru.Cache{MaxEntries: maxEntries}}
+}
+
+// Get tries to get CRL from cache, returns error if failed
+func (c *LRUCRLCache) Get(key string) (*CRLCacheItem, error) {
+	log.Debugf("CRL: LRU cache: loading '%s'", key)
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	value, ok := c.cache.Get(key)
+	if ok {
+		value, _ := value.(*CRLCacheItem)
+		log.Debugf("CRL: LRU cache: '%s' found", key)
+		return value, nil
+	}
+
+	return nil, ErrCacheKeyNotFound
+}
+
+// Put stores CRL in cache
+func (c *LRUCRLCache) Put(key string, value *CRLCacheItem) error {
+	c.mutex.Lock()
+	c.cache.Add(key, value)
+	c.mutex.Unlock()
+	log.Debugf("CRL: LRU cache: inserted '%s'", key)
+	return nil
+}
+
+// Remove removes item from cache
+func (c *LRUCRLCache) Remove(key string) error {
+	c.mutex.Lock()
+	c.cache.Remove(key)
+	c.mutex.Unlock()
+	log.Debugf("CRL: LRU cache: removed '%s'", key)
+	return nil
+}
+
+// DefaultCRLVerifier is a default implementation of CRLVerifier
+type DefaultCRLVerifier struct {
+	Config CRLConfig
+	Client CRLClient
+	Cache  CRLCache
+}
+
+// Tries to find cached CRL, fetches using v.Client if not found, checks the signature of CRL using issuerCert
+func (v DefaultCRLVerifier) getCachedOrFetch(uri string, issuerCert *x509.Certificate) (*pkix.CertificateList, error) {
+	// Try v.Cache first, but only if caching is enabled (cache time > 0)
+	if v.Config.cacheTime > 0 {
+		cacheItem, err := v.Cache.Get(v.Config.uri)
+		if cacheItem != nil {
+			if err != nil {
+				// non-empty result + error, should never happen
+				return nil, err
+			}
+
+			if time.Now().Before(cacheItem.Fetched.Add(v.Config.cacheTime)) {
+				return &cacheItem.CRL, nil
+			}
+		}
+	}
+
+	// Not found in cache (or the CRL was outdated), gotta fetch
+	rawCRL, err := v.Client.Fetch(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	crl, err := x509.ParseCRL(rawCRL)
+	if err != nil {
+		log.WithError(err).Debugf("CRL: Cannot parse CRL from '%s'", uri)
+		return nil, err
+	}
+
+	err = issuerCert.CheckCRLSignature(crl)
+	if err != nil {
+		log.WithError(err).Warnf("CRL: Failed to check signature for CRL at %s", uri)
+		return nil, err
+	}
+
+	if crl.TBSCertList.NextUpdate.Before(time.Now()) {
+		log.Warnf("CRL: CRL at %s is outdated", uri)
+		return nil, ErrOutdatedCRL
+	}
+
+	if v.Config.cacheTime > 0 {
+		cacheItem := &CRLCacheItem{Fetched: time.Now(), CRL: *crl}
+		v.Cache.Put(uri, cacheItem)
+	}
+
+	return crl, nil
+}
+
+func (v DefaultCRLVerifier) verifyCertWithIssuer(cert, issuer *x509.Certificate) error {
+	log.Debugf("CRL: Verifying '%s'", cert.Subject.String())
+
+	if len(v.Config.uri) == 0 && v.Config.fromCert == crlFromCertIgnore {
+		log.Debugln("CRL: Skipping check since no config URI specified and we were told to ignore URIs from certificate")
+		return nil
+	}
+
+	for _, crlDistributionPoint := range cert.CRLDistributionPoints {
```

Comment:
Not-so-unnecessary since using using CRL URLs from cert may be disabled in config, but in this loop we'll still be able to see debug log with contained URLs. No more than for debugging purposes